### ISSUE TITLE
HUGE speedup opportunity with many files

### ISSUE THREAD

**Issue by jermerf** [syncthing/syncthing#9273]
I have been watching for the last hour as Syncthing slowly transfers a massive 18MB worth of files, of which there are 6311 files. And yes it's on TCP LAN and I'm sitting beside the router with both devices.

Instead of waiting:
- DEV A: Zipped the 6311 files
- DEV A: Moved the folder with 6311 files out of the folder Syncthing was watching
- DEV B: Stop Syncthing
- Delete the folder with some of the 6311 files
- DEV B: Start Syncthing
- Waited for the transfer of the one zip file over literally 2 seconds
- DEV B: Unzipped the files
- DEV A: Moved the folder back on the original device
- Delete the Zip files
- All done in about a minute

If there are many files to transfer, they can be zipped and transferred and then unzipped. It is WAY faster than waiting as it slowly counts files. This could be made into a process to speed things up significantly. Because in no world does it help me out to transfer at less than 18MB/Hour.

**Comment by AudriusButkevicius** [comment]
There is quite a large overhead per file, due to various things, but things are the way they are, mostly to ensure no data loss.

Namely every file needs to be recorded in the database before it's put in it's final place, each file needs to be fsynced, this adds potentially 2-3 second overhead per file.

In your case, what do you think will happen if your computer crashes half way through unzipping? It will produce half unzipped files, which will look like "newer files" and overwrite the proper files on the other side, leading to data loss.

I think as it stands I think the choices we've made are the right ones, as it has strong guarantees around not losing data, at the expense of speed.

You can probably achieve similar results by disabling fsync, and accepting data loss in case you lose power/have the machine crash.

**Comment by jermerf** [comment]
I have to say I'm disappointed this was closed with so little discussion. I can't believe that it is the decision of the entire group of contributors that it is not worth it to achieve a 100x+ speed boost. End of discussion. I understand completely the importance of the integrity of the data, but at what cost? If I weren't already using Syncthing, I would probably not have started using it. Because unless you are using few large files, I would say it is unusable. It's just too slow.

I wrote a system in NodeJS a while back that just does MD5 Checksums to see if files are different, and switched from that to Syncthing. And it was WAY faster. I don't understand this 2-3 second overhead/per file, when my md5 Frankenstein which used SQLite could stick in dozens if not hundreds of files per second.

Why close this issue? The performance of Syncthing with many files - IS EMPIRACALLY SLOW AND BAD! And can be MUCH better. I have done better, but this was supposed to be better. I guess not.
